Added `gen_substrs` support

ewynx opened the pull request:
Extraction of 1 or more substrings can be requested in both settings (raw, decomposed):
- raw: pass on a json file with `-s <SUBSTRS_JSON_PATH>` that contains the state transitions that should be revealed
- decomposed: mark the part as public

This information is added to `substrings` in the `RegexAndDFA`. In the case of raw only `substring_ranges` is filled so this is what this implementation uses.

## How it works
The adjustments when `gen_substrs = true` are:
- return type of `regex_match` becomes an array length of #substrs, each of which is a BoundedVec. We don't know the substring length beforehand, but it is bounded by N
- fill #substrs of BoundedVec with the bytes when in the corresponding state

If the bool is false, nothing changes. 

Note: For the raw setting the boolean was by default set to true, I changed it to false, just like in the decomposed setting. Asked in the TG group if this is ok. 

## Testing

### Decomposed

Create a file called `substring_test.json` containing:
```
{
  "parts":[
      {
          "is_public": false,
          "regex_def": "email was meant for @"
      },
      {
          "is_public": true,
          "regex_def": "[a-z]+"
      },
      {
          "is_public": false,
          "regex_def": "\\. And the next substring:"
      },
      {
          "is_public": true,
          "regex_def": "[a-z]+"
      }
  ]
}
```
Run
```
cargo run --bin zk-regex decomposed -d substring_test.json --noir-file-path testfile.nr -g true
```

### Raw

This is the example from the README.md; create `./simple_regex_substrs.json` containing:
```
    {
        "transitions": [
            [
                [
                    2,
                    3
                ]
            ],
            [
                [
                    6,
                    7
                ],
                [
                    7,
                    7
                ]
            ],
            [
                [
                    8,
                    9
                ]
            ]
        ]
    }
```
Run
```
cargo run --bin zk-regex raw -r "1=(a|b) (2=(b|c)+ )+d" -s ./simple_regex_substrs.json --noir-file-path simple_regex_substrs.nr -g true`

```

Note that the second substring is marked by 2 transitions [6,7] and [7,7], which is needed for the circom impl. However, in Noir we only need to know the substring is in state 7, so this is the information that is taken from `substring_ranges`.

ewynx committed:
Added gen_substrs support for Noir. This works both in decomposed & raw setting.

The substrings are returned as BoundedVec since we don't know their exact length upfront, but we know they're not longer than N.
To support both settings (decomposed and raw) we have to use `substring_ranges` instead of `substring_boundaries`.

ewynx committed:
Set gen_substrs to false by default for raw.

olehmisar commented:
this changes a lot. We need a testing suite to reliably verify that it works. Any idea how to test noir support? I believe we can use Noir wasm compiler for that
https://github.com/noir-lang/noir-starter/blob/3533cdc2e167bce5472a45a2fa5e9e053cea3459/vite-hardhat/circuit/compile.ts

ewynx commented:
I agree, good testing is needed. We've been working on a testing suite [here](https://github.com/hashcloak/zk-regex-tests-and-examples). This generates the Noir code using the compiler and then verifies the working of that code against different inputs. However, it is a WIP, so we'll let you know when the testing for `gen_substrs` is included. 

For the implementation I did manual testing in the 2 ways mentioned above.

olehmisar commented:
@ewynx ok, I trust you here. If this PR is a blocker, I can merge it. Please open an issue to add tests for gen_strings before the merge.

ewynx commented:
I'll have to fix the following bug: the nr of substrings is not necessarily known beforehand. 
For example "(substr=(a|b)+ )+" where the substring to extract is (a|b). Valid inputs are:

"substr=a substr=b" => extracts 2 substrings, namely "a" and "b"
"substr=b" => extracts 1 substring, namely "b"

ewynx committed:
Per state 1 or more substrings will be extracted, depending on the regex and input. This fix makes sure this is supported.

Changes:
- regex_match returns a Vec of substrings instead of an array with known length
- per state where substrings have to be extracted; add the byte either to a new substring or an already started one

Note that substr_count is used to extract the correct "current" substring from the Vec. This is a workaround - first implementation was using `pop` but this gave an error.

ewynx commented:
Above mentioned task is done ✅
